library/core/src/ptr/mod.rs

```diff
@@ -1624,8 +1630,13 @@ pub(crate) unsafe fn align_offset<T: Sized>(p: *const T, a: usize) -> usize {
         }
     }
 
-    let addr = p.addr();
     let stride = mem::size_of::<T>();
+
+    // SAFETY: At runtime transmuting a pointer to `usize` is always safe, because they have the
+    // same layout. During const eval we hook this function to ensure that the pointer always has
+    // an address (only the standard library can do this).
```

Comment:
that's... not quite true. Even at runtime those transmutes are fishy and until there's a decision for allowing such transmutes we should avoid them.

Maybe keep this function body non-const and make the mir interpreter hook do the logic for the integral value, too?